--- LLParser.cpp.orig	2011-11-28 10:00:38.008628498 +1000
+++ LLParser.cpp	2011-11-28 10:13:24.120606475 +1000
@@ -1261,8 +1261,14 @@
     // If the type hasn't been defined yet, create a forward definition and
     // remember where that forward def'n was seen (in case it never is defined).
     if (Entry.first == 0) {
-      Entry.first = StructType::create(Context, Lex.getStrVal());
-      Entry.second = Lex.getLoc();
+        // this here for extempore
+  	if (M->getTypeByName(Lex.getStrVal())) {
+           Entry.first = M->getTypeByName(Lex.getStrVal());
+           Entry.second = SMLoc();
+        } else {	
+      	   Entry.first = StructType::create(Context, Lex.getStrVal());
+           Entry.second = Lex.getLoc();
+	}
     }
     Result = Entry.first;
     Lex.Lex();
